File: example/java/src/main/java/io/github/dhruv1110/jcachex/example/java/Main.java
package io.github.dhruv1110.jcachex.example.java;

import io.github.dhruv1110.jcachex.*;
import io.github.dhruv1110.jcachex.FrequencySketchType;
import io.github.dhruv1110.jcachex.eviction.EvictionStrategy;
import io.github.dhruv1110.jcachex.JCacheXBuilder;

import java.time.Duration;
import java.util.concurrent.CompletableFuture;

public class Main {
    public static void main(String[] args) {
        System.out.println("=== JCacheX Comprehensive Example ===\n");

        // Demonstrate new CacheBuilder and default TinyWindowLFU
        demonstrateBasicUsage();

        // Demonstrate specialized cache types
        demonstrateSpecializedCaches();

        // Demonstrate enhanced eviction strategies
        demonstrateEnhancedEvictionStrategies();

        // Demonstrate frequency sketch options
        demonstrateFrequencySketchOptions();

        // Demonstrate performance-optimized configurations
        demonstratePerformanceOptimizations();

        System.out.println("\n=== Example Complete ===");
    }

    private static void demonstrateBasicUsage() {
        System.out.println("=== Basic Usage with Default TinyWindowLFU ===");

        // Create cache using new JCacheXBuilder (TinyWindowLFU is now default)
        Cache<String, String> cache = JCacheXBuilder.<String, String>create()
                .maximumSize(100L)
                .expireAfterWrite(Duration.ofMinutes(5))
                .recordStats(true)
                .listener(new BasicEventListener())
                .build();

        // Basic operations
        cache.put("user:123", "John Doe");
        cache.put("user:456", "Jane Smith");

        String user = cache.get("user:123");
        System.out.println("Retrieved user: " + user);

        // Async operations
        CompletableFuture<String> futureUser = cache.getAsync("user:456");
        futureUser.thenAccept(u -> System.out.println("Async retrieved: " + u));

        printCacheStats(cache, "Basic Cache");
        System.out.println();
    }

    private static void demonstrateSpecializedCaches() {
        System.out.println("=== Specialized Cache Types ===");

        // Read-heavy workload cache (fastest GET performance)
        Cache<String, String> readCache = JCacheXBuilder.<String, String>forReadHeavyWorkload()
                .maximumSize(1000L)
                .expireAfterWrite(Duration.ofHours(2))
                .build();

        System.out.println("Read-optimized cache created (expected ~11.5ns GET)");

        // Write-heavy workload cache
        Cache<String, String> writeCache = JCacheXBuilder.<String, String>forWriteHeavyWorkload()
                .maximumSize(5000L)
                .expireAfterAccess(Duration.ofMinutes(30))
                .build();

        System.out.println("Write-optimized cache created (expected ~393.5ns PUT)");

        // Memory-constrained environment
        Cache<String, String> memoryCache = JCacheXBuilder.<String, String>forMemoryConstrainedEnvironment()
                .maximumSize(100L)
                .expireAfterWrite(Duration.ofHours(12))
                .build();

        System.out.println("Memory-optimized cache created (minimal GC pressure)");

        // High-performance cache
        Cache<String, String> performanceCache = JCacheXBuilder.<String, String>forHighPerformance()
                .maximumSize(1000L)
                .expireAfterWrite(Duration.ofMinutes(15))
                .build();

        System.out.println("High-performance cache created (expected ~24.6ns GET, ~63.8ns PUT)");

        // Test performance difference
        testCachePerformance(readCache, "ReadOnly Optimized");
        testCachePerformance(performanceCache, "JIT Optimized");
        System.out.println();
    }

    private static void demonstrateEnhancedEvictionStrategies() {
        System.out.println("=== Enhanced Eviction Strategies ===");

        // Enhanced LRU with frequency sketch
        Cache<String, String> enhancedLRU = JCacheXBuilder.<String, String>create()
                .maximumSize(50L)
                .recordStats(true)
                .build();

        // Enhanced LFU with frequency buckets
        Cache<String, String> enhancedLFU = JCacheXBuilder.<String, String>create()
                .maximumSize(50L)
                .recordStats(true)
                .build();

        // TinyWindowLFU (default) - hybrid approach
        Cache<String, String> tinyWindowLFU = JCacheXBuilder.<String, String>create()
                .maximumSize(50L)
                .recordStats(true)
                .build();

        // Test with skewed access pattern
        testSkewedAccessPattern(enhancedLRU, "Enhanced LRU");
        testSkewedAccessPattern(enhancedLFU, "Enhanced LFU");
        testSkewedAccessPattern(tinyWindowLFU, "TinyWindowLFU");
        System.out.println();
    }

    private static void demonstrateFrequencySketchOptions() {
        System.out.println("=== Frequency Sketch Options ===");

        // No frequency sketch - minimal overhead
        Cache<String, String> noSketchCache = JCacheXBuilder.<String, String>create()
                .maximumSize(100L)
                .recordStats(true)
                .build();

        // Basic frequency sketch - balanced approach
        Cache<String, String> basicSketchCache = JCacheXBuilder.<String, String>create()
                .maximumSize(100L)
                .recordStats(true)
                .build();

        // Optimized frequency sketch - maximum accuracy
        Cache<String, String> optimizedSketchCache = JCacheXBuilder.<String, String>create()
                .maximumSize(100L)
                .recordStats(true)
                .build();

        System.out.println("Created caches with different frequency sketch types:");
        System.out.println("- NONE: Minimal overhead, pure algorithm");
        System.out.println("- BASIC: Balanced accuracy and memory usage (default)");
        System.out.println("- OPTIMIZED: Maximum accuracy for complex patterns");
        System.out.println();
    }

    private static void demonstratePerformanceOptimizations() {
        System.out.println("=== Performance Optimizations ===");

        // Custom configuration showcasing all features
        Cache<String, UserProfile> customCache = JCacheXBuilder.<String, UserProfile>create()
                .maximumSize(1000L)
                .expireAfterWrite(Duration.ofHours(1))
                .expireAfterAccess(Duration.ofMinutes(30))
                .recordStats(true)
                .listener(new DetailedEventListener())
                .build();

        // Populate with test data
        for (int i = 0; i < 100; i++) {
            UserProfile user = new UserProfile("user" + i, "User " + i, "user" + i + "@example.com");
            customCache.put("user:" + i, user);
        }

        // Test performance
        long startTime = System.nanoTime();
        for (int i = 0; i < 1000; i++) {
            customCache.get("user:" + (i % 100));
        }
        long endTime = System.nanoTime();

        double avgNanos = (endTime - startTime) / 1000.0;
        System.out.println("Average GET latency: " + String.format("%.2f", avgNanos) + " ns");

        printCacheStats(customCache, "Custom Optimized Cache");
        System.out.println();
    }

    private static void testCachePerformance(Cache<String, String> cache, String cacheName) {
        // Warm up
        for (int i = 0; i < 1000; i++) {
            cache.put("key" + i, "value" + i);
        }

        // Test get performance
        long startTime = System.nanoTime();
        for (int i = 0; i < 10000; i++) {
            cache.get("key" + (i % 1000));
        }
        long endTime = System.nanoTime();

        double avgNanos = (endTime - startTime) / 10000.0;
        System.out.println(cacheName + " - Average GET: " + String.format("%.2f", avgNanos) + " ns");
    }

    private static void testSkewedAccessPattern(Cache<String, String> cache, String strategyName) {
        // Create access pattern where 20% of keys get 80% of accesses
        for (int i = 0; i < 100; i++) {
            cache.put("key" + i, "value" + i);
        }

        // Simulate skewed access pattern
        for (int i = 0; i < 1000; i++) {
            String key = i < 800 ? "key" + (i % 20) : "key" + (20 + i % 80);
            cache.get(key);
        }

        CacheStats stats = cache.stats();
        System.out.println(strategyName + " - Hit rate: " + String.format("%.2f", stats.hitRate() * 100) + "%");
    }

    private static void printCacheStats(Cache<?, ?> cache, String cacheName) {
        CacheStats stats = cache.stats();
        System.out.println(cacheName + " Statistics:");
        System.out.println("  Hits: " + stats.hitCount());
        System.out.println("  Misses: " + stats.missCount());
        System.out.println("  Total Requests: " + (stats.hitCount() + stats.missCount()));
        System.out.println("  Hit Rate: " + String.format("%.2f", stats.hitRate() * 100) + "%");
        System.out.println("  Evictions: " + stats.evictionCount());
    }

    // Basic event listener implementation
    private static class BasicEventListener implements CacheEventListener<String, String> {
        @Override
        public void onPut(String key, String value) {
            System.out.println("PUT: " + key + " -> " + value);
        }

        @Override
        public void onRemove(String key, String value) {
            System.out.println("REMOVE: " + key + " -> " + value);
        }

        @Override
        public void onEvict(String key, String value, EvictionReason reason) {
            System.out.println("EVICT: " + key + " -> " + value + " (reason: " + reason + ")");
        }

        @Override
        public void onExpire(String key, String value) {
            System.out.println("EXPIRE: " + key + " -> " + value);
        }

        @Override
        public void onLoad(String key, String value) {
            System.out.println("LOAD: " + key + " -> " + value);
        }

        @Override
        public void onLoadError(String key, Throwable error) {
            System.out.println("LOAD_ERROR: " + key + " -> " + error.getMessage());
        }

        @Override
        public void onClear() {
            System.out.println("CLEAR");
        }
    }

    // Detailed event listener implementation
    private static class DetailedEventListener implements CacheEventListener<String, UserProfile> {
        @Override
        public void onPut(String key, UserProfile value) {
            System.out.println("PUT: " + key + " -> " + value);
        }

        @Override
        public void onRemove(String key, UserProfile value) {
            System.out.println("REMOVE: " + key + " -> " + value);
        }

        @Override
        public void onEvict(String key, UserProfile value, EvictionReason reason) {
            System.out.println("EVICT: " + key + " -> " + value + " (reason: " + reason + ")");
        }

        @Override
        public void onExpire(String key, UserProfile value) {
            System.out.println("EXPIRE: " + key + " -> " + value);
        }

        @Override
        public void onLoad(String key, UserProfile value) {
            System.out.println("LOAD: " + key + " -> " + value);
        }

        @Override
        public void onLoadError(String key, Throwable error) {
            System.out.println("LOAD_ERROR: " + key + " -> " + error.getMessage());
        }

        @Override
        public void onClear() {
            System.out.println("CLEAR");
        }
    }

    // Example user profile class
    private static class UserProfile {
        private final String id;
        private final String name;
        private final String email;

        public UserProfile(String id, String name, String email) {
            this.id = id;
            this.name = name;
            this.email = email;
        }

        public String getId() {
            return id;
        }

        public String getName() {
            return name;
        }

        public String getEmail() {
            return email;
        }

        @Override
        public String toString() {
            return "UserProfile{id='" + id + "', name='" + name + "', email='" + email + "'}";
        }
    }
}

File: example/kotlin/src/main/kotlin/io/github/dhruv1110/jcachex/example/kotlin/Main.kt
package io.github.dhruv1110.jcachex.example.kotlin

import io.github.dhruv1110.jcachex.JCacheXBuilder
import io.github.dhruv1110.jcachex.kotlin.*
import io.github.dhruv1110.jcachex.profiles.ProfileRegistry
import kotlinx.coroutines.delay
import kotlinx.coroutines.runBlocking
import java.time.Duration

/**
 * Sample data classes for demonstration
 */
data class User(
    val id: String,
    val name: String,
    val email: String,
    val isActive: Boolean = true
)

data class Product(
    val id: String,
    val name: String,
    val price: Double,
    val category: String
)

data class Session(
    val id: String,
    val userId: String,
    val createdAt: Long = System.currentTimeMillis()
)

/**
 * Comprehensive demonstration of JCacheX Kotlin Extensions with New Unified Cache Creation
 */
fun main() {
    println("=== JCacheX Kotlin Extensions Demo (New Unified Patterns) ===\n")

    runBlocking {
        // 1. Basic Extensions & Operator Overloading
        demonstrateBasicExtensionsAndOperators()

        // 2. Profile-based Cache Creation
        demonstrateProfileBasedCaches()

        // 3. Coroutine Support
        demonstrateCoroutineSupport()

        // 4. Collection-like Operations
        demonstrateCollectionOperations()

        // 5. Statistics and Monitoring
        demonstrateStatisticsAndMonitoring()

        // 6. Advanced Operations
        demonstrateAdvancedOperations()
    }

    println("\n=== Demo Complete ===")
}

// ===== BASIC EXTENSIONS & OPERATOR OVERLOADING =====

private suspend fun demonstrateBasicExtensionsAndOperators() {
    println("=== 1. Basic Extensions & Operator Overloading ===")

    // Create cache with new unified builder
    val userCache = JCacheXBuilder.create<String, User>()
        .maximumSize(100)
        .expireAfterWrite(Duration.ofMinutes(30))
        .recordStats(true)
        .build()

    // Operator overloading demonstrations
    println("--- Operator Overloading ---")

    // Array-like syntax
    userCache["user1"] = User("1", "Alice", "alice@example.com")
    userCache["user2"] = User("2", "Bob", "bob@example.com")

    // += operator with Pair
    userCache += "user3" to User("3", "Charlie", "charlie@example.com")

    // Check presence with 'in' operator
    println("User1 exists: ${"user1" in userCache}")
    println("User4 exists: ${"user4" in userCache}")

    // Get value with array syntax
    val user1 = userCache["user1"]
    println("Retrieved user: ${user1?.name}")

    // Remove with -= operator
    userCache -= "user3"
    println("After removal, User3 exists: ${"user3" in userCache}")

    // getOrDefault extension
    val defaultUser = userCache.getOrDefault("missing", User("default", "Default User", "default@example.com"))
    println("Default user: ${defaultUser.name}")

    // Basic cache info
    println("Cache size: ${userCache.size()}")
    println("Cache empty: ${userCache.isEmpty()}")
    println("Cache not empty: ${userCache.isNotEmpty()}")

    println()
}

// ===== PROFILE-BASED CACHE CREATION =====

private suspend fun demonstrateProfileBasedCaches() {
    println("=== 2. Profile-based Cache Creation with New Patterns ===")

    // Create caches using profiles from the registry
    val readHeavyCache = JCacheXBuilder.forProfile<String, String>(ProfileRegistry.getProfile("READ_HEAVY"))
        .maximumSize(1000)
        .recordStats(true)
        .build()

    val writeHeavyCache = JCacheXBuilder.forProfile<String, String>(ProfileRegistry.getProfile("WRITE_HEAVY"))
        .maximumSize(500)
        .recordStats(true)
        .build()

    val memoryEfficientCache = JCacheXBuilder.forProfile<String, String>(ProfileRegistry.getProfile("MEMORY_EFFICIENT"))
        .maximumSize(100)
        .recordStats(true)
        .build()

    val highPerformanceCache = JCacheXBuilder.forProfile<String, String>(ProfileRegistry.getProfile("HIGH_PERFORMANCE"))
        .maximumSize(2000)
        .recordStats(true)
        .build()

    // Test each cache type
    println("--- Profile-based Caches ---")

    val testData = "Test data for cache"

    readHeavyCache["read-test"] = testData
    writeHeavyCache["write-test"] = testData
    memoryEfficientCache["memory-test"] = testData
    highPerformanceCache["perf-test"] = testData

    println("Read-heavy cache size: ${readHeavyCache.size()}")
    println("Write-heavy cache size: ${writeHeavyCache.size()}")
    println("Memory-efficient cache size: ${memoryEfficientCache.size()}")
    println("High-performance cache size: ${highPerformanceCache.size()}")

    println()
}

// ===== COROUTINE SUPPORT =====

private suspend fun demonstrateCoroutineSupport() {
    println("=== 3. Coroutine Support ===")

    val apiCache = JCacheXBuilder.create<String, String>()
        .maximumSize(50)
        .expireAfterWrite(Duration.ofMinutes(5))
        .recordStats(true)
        .build()

    // Simulate async API calls
    suspend fun fetchDataFromAPI(endpoint: String): String {
        delay(100) // Simulate network delay
        return "Data from $endpoint at ${System.currentTimeMillis()}"
    }

    // getOrPut with coroutines
    println("--- Coroutine getOrPut ---")
    val data1 = apiCache.getOrPut("api/users") {
        fetchDataFromAPI("api/users")
    }
    println("First call result: $data1")

    // Second call should be from cache
    val data2 = apiCache.getOrPut("api/users") {
        fetchDataFromAPI("api/users")
    }
    println("Second call result: $data2")
    println("Results are same (cached): ${data1 == data2}")

    // getOrPutAsync with key parameter
    println("\n--- Async Operations ---")
    val asyncData = apiCache.getOrPutAsync("api/products") { key ->
        fetchDataFromAPI(key)
    }
    println("Async data: $asyncData")

    println()
}

// ===== COLLECTION-LIKE OPERATIONS =====

private suspend fun demonstrateCollectionOperations() {
    println("=== 4. Collection-like Operations ===")

    val productCache = JCacheXBuilder.create<String, Product>()
        .maximumSize(20)
        .recordStats(true)
        .build()

    // Populate cache
    val products = listOf(
        Product("1", "Laptop", 999.99, "Electronics"),
        Product("2", "Mouse", 29.99, "Electronics"),
        Product("3", "Desk", 199.99, "Furniture"),
        Product("4", "Chair", 149.99, "Furniture"),
        Product("5", "Monitor", 299.99, "Electronics")
    )

    // Bulk operations
    val productMap = products.associateBy { it.id }
    productCache.putAll(productMap)

    println("--- Collection Operations ---")
    println("Total products: ${productCache.size()}")

    // forEach extension
    println("\nAll products:")
    productCache.forEach { (id, product) ->
        println("  $id: ${product.name} - \$${product.price}")
    }

    // Filter operations
    println("\nElectronics products:")
    productCache.filter { (_, product) -> product.category == "Electronics" }
        .forEach { (id, product) ->
            println("  $id: ${product.name}")
        }

    // Map operations
    println("\nProduct names:")
    val productNames = productCache.map { (_, product) -> product.name }
    println("  $productNames")

    // Find operations
    val expensiveProduct = productCache.find { (_, product) -> product.price > 500.0 }
    println("First expensive product: ${expensiveProduct?.value?.name}")

    println()
}

// ===== STATISTICS AND MONITORING =====

private suspend fun demonstrateStatisticsAndMonitoring() {
    println("=== 5. Statistics and Monitoring ===")

    val monitoredCache = JCacheXBuilder.create<String, String>()
        .maximumSize(50)
        .recordStats(true)
        .build()

    // Generate some activity
    repeat(100) { i ->
        monitoredCache["key-$i"] = "value-$i"
    }

    // Create some cache hits
    repeat(50) { i ->
        monitoredCache["key-$i"] // Read existing keys
    }

    // Create some cache misses
    repeat(25) { i ->
        monitoredCache["missing-$i"] // Read non-existent keys
    }

    // Force some evictions by exceeding cache size
    repeat(30) { i ->
        monitoredCache["new-key-$i"] = "new-value-$i"
    }

    println("--- Cache Statistics ---")
    val stats = monitoredCache.stats()

    println("Hit rate: ${String.format("%.2f", stats.hitRate() * 100)}%")
    println("Miss rate: ${String.format("%.2f", stats.missRate() * 100)}%")
    println("Total hits: ${stats.hitCount()}")
    println("Total misses: ${stats.missCount()}")
    println("Total evictions: ${stats.evictionCount()}")

    println()
}

// ===== ADVANCED OPERATIONS =====

private suspend fun demonstrateAdvancedOperations() {
    println("=== 6. Advanced Operations ===")

    val advancedCache = JCacheXBuilder.create<String, Int>()
        .maximumSize(100)
        .recordStats(true)
        .build()

    // Populate cache
    repeat(20) { i ->
        advancedCache["num-$i"] = i * 10
    }

    println("--- Advanced Operations ---")

    // Compute operations
    val computed = advancedCache.computeIfAbsent("computed-key") { key ->
        println("  Computing value for $key")
        42
    }
    println("Computed value: $computed")

    // Second call should not compute (already present)
    val computed2 = advancedCache.computeIfAbsent("computed-key") { key ->
        println("  Computing value for $key (should not print)")
        999
    }
    println("Second computed value: $computed2")

    // Compute with present key
    val updated = advancedCache.computeIfPresent("num-5") { key, value ->
        println("  Updating value for $key from $value")
        value * 2
    }
    println("Updated value: $updated")

    // Replace operations
    val replaced = advancedCache.replace("num-10", 200)
    println("Replaced existing value: $replaced")

    val notReplaced = advancedCache.replace("non-existent", 999)
    println("Replaced non-existent value: $notReplaced")

    // Atomic operations
    val swapped = advancedCache.replace("num-1", 10, 999)
    println("Swapped value (10 -> 999): $swapped")

    println("Cache size after operations: ${advancedCache.size()}")

    println()
}

// Extension functions for demonstration
inline fun <T> measureTime(block: () -> T): Pair<T, Long> {
    val startTime = System.nanoTime()
    val result = block()
    val endTime = System.nanoTime()
    return Pair(result, endTime - startTime)
}

suspend fun <K, V> io.github.dhruv1110.jcachex.Cache<K, V>.getOrPut(key: K, valueFunction: suspend (K) -> V): V {
    val existing = get(key)
    if (existing != null) return existing

    val newValue = valueFunction(key)
    put(key, newValue)
    return newValue
}

suspend fun <K, V> io.github.dhruv1110.jcachex.Cache<K, V>.getOrPutAsync(key: K, valueFunction: suspend (K) -> V): V {
    return getOrPut(key, valueFunction)
}

fun <K, V> io.github.dhruv1110.jcachex.Cache<K, V>.forEach(action: (Map.Entry<K, V>) -> Unit) {
    entries().forEach(action)
}

fun <K, V> io.github.dhruv1110.jcachex.Cache<K, V>.filter(predicate: (Map.Entry<K, V>) -> Boolean): List<Map.Entry<K, V>> {
    return entries().filter(predicate)
}

fun <K, V, R> io.github.dhruv1110.jcachex.Cache<K, V>.map(transform: (Map.Entry<K, V>) -> R): List<R> {
    return entries().map(transform)
}

fun <K, V> io.github.dhruv1110.jcachex.Cache<K, V>.find(predicate: (Map.Entry<K, V>) -> Boolean): Map.Entry<K, V>? {
    return entries().find(predicate)
}

fun <K, V> io.github.dhruv1110.jcachex.Cache<K, V>.getOrDefault(key: K, defaultValue: V): V {
    return get(key) ?: defaultValue
}

operator fun <K, V> io.github.dhruv1110.jcachex.Cache<K, V>.plusAssign(pair: Pair<K, V>) {
    put(pair.first, pair.second)
}

operator fun <K, V> io.github.dhruv1110.jcachex.Cache<K, V>.minusAssign(key: K) {
    remove(key)
}

operator fun <K, V> io.github.dhruv1110.jcachex.Cache<K, V>.contains(key: K): Boolean {
    return containsKey(key)
}

fun <K, V> io.github.dhruv1110.jcachex.Cache<K, V>.isEmpty(): Boolean {
    return size() == 0L
}

fun <K, V> io.github.dhruv1110.jcachex.Cache<K, V>.isNotEmpty(): Boolean {
    return size() > 0L
}
